File: apps/rekky_backend/src/migrate.ts
import 'dotenv/config';
import { readFile, readdir } from 'node:fs/promises';
import { fileURLToPath } from 'node:url';
import path from 'node:path';
import { createPool } from './db.js';

const databaseUrl = process.env.DATABASE_URL;
if (!databaseUrl) throw new Error('DATABASE_URL is required');
const pool = createPool(databaseUrl);
const directory = fileURLToPath(new URL('../migrations/', import.meta.url));
try {
  const files = (await readdir(directory)).filter((file) => /^\d+_[\w-]+\.sql$/.test(file)).sort();
  const client = await pool.connect();
  try {
    await client.query('BEGIN');
    await client.query('SELECT pg_advisory_xact_lock(732781)');
    await client.query('CREATE TABLE IF NOT EXISTS schema_migrations (name text PRIMARY KEY, applied_at timestamptz NOT NULL DEFAULT now())');
    for (const file of files) {
      const exists = await client.query('SELECT 1 FROM schema_migrations WHERE name = $1', [file]);
      if (exists.rowCount) continue;
      await client.query(await readFile(path.join(directory, file), 'utf8'));
      await client.query('INSERT INTO schema_migrations(name) VALUES ($1)', [file]);
      process.stdout.write(`Applied ${file}\n`);
    }
    await client.query('COMMIT');
  } catch (error) {
    await client.query('ROLLBACK');
    throw error;
  } finally {
    client.release();
  }
} finally {
  await pool.end();
}

File: apps/rekky_backend/src/migrate.rs
use sqlx::PgPool;

const MIGRATIONS: &[(&str, &str)] = &[
    (
        "001_foundation.sql",
        include_str!("../migrations/001_foundation.sql"),
    ),
    (
        "002_owner_consistency.sql",
        include_str!("../migrations/002_owner_consistency.sql"),
    ),
    (
        "003_processing_permission.sql",
        include_str!("../migrations/003_processing_permission.sql"),
    ),
    (
        "004_voice_transcription.sql",
        include_str!("../migrations/004_voice_transcription.sql"),
    ),
    (
        "005_voice_attempt_cap.sql",
        include_str!("../migrations/005_voice_attempt_cap.sql"),
    ),
    (
        "006_transcript_extraction.sql",
        include_str!("../migrations/006_transcript_extraction.sql"),
    ),
    (
        "007_auto_voice_processing.sql",
        include_str!("../migrations/007_auto_voice_processing.sql"),
    ),
    (
        "008_structured_recommendations.sql",
        include_str!("../migrations/008_structured_recommendations.sql"),
    ),
    (
        "009_category_search.sql",
        include_str!("../migrations/009_category_search.sql"),
    ),
    (
        "010_readable_source.sql",
        include_str!("../migrations/010_readable_source.sql"),
    ),
];

pub async fn run(pool: &PgPool) -> Result<(), sqlx::Error> {
    let mut transaction = pool.begin().await?;
    sqlx::query("SELECT pg_advisory_xact_lock(732781)")
        .execute(&mut *transaction)
        .await?;
    sqlx::query("CREATE TABLE IF NOT EXISTS schema_migrations (name text PRIMARY KEY, applied_at timestamptz NOT NULL DEFAULT now())")
        .execute(&mut *transaction).await?;
    for (name, script) in MIGRATIONS {
        let exists = sqlx::query("SELECT 1 FROM schema_migrations WHERE name=$1")
            .bind(name)
            .fetch_optional(&mut *transaction)
            .await?
            .is_some();
        if !exists {
            sqlx::raw_sql(*script).execute(&mut *transaction).await?;
            sqlx::query("INSERT INTO schema_migrations(name) VALUES ($1)")
                .bind(name)
                .execute(&mut *transaction)
                .await?;
            println!("Applied {name}");
        }
    }
    transaction.commit().await?;
    Ok(())
}
